ckanext/dia/commands.py

```diff
@@ -68,38 +70,72 @@ def cleanup_datastore_once(self):
 
         # query datastore to get all resources from the _table_metadata
         resource_id_list = []
-        for offset in itertools.count(start=0, step=100):
+        total_tested = 0
+        for offset in itertools.count(start=start_offset, step=100):
             print "Load metadata records from datastore (offset: %s)" % offset
-            record_list, has_next_page = self._get_datastore_table_page(context, offset)  # noqa
+            record_list, tested = self._get_datastore_table_page(context, offset)  # noqa
+            total_tested += tested
             resource_id_list.extend(record_list)
-            if not has_next_page:
+            if not tested > 0:
                 break
             if len(resource_id_list) > 250:
                 # Run a small chunk of the dataset to avoid locking up the
                 # database for a *really* long time(read: until postgres is
                 # restarted)
                 break
 
+        print('Total resources missing but referenced in datastore: {}'.format(len(resource_id_list)))
+
+        def datastore_record_count(context, resource_id):
+            """
+            Wraps the search so that it returns None if NotFound, otherwise the total
+            """
+            try:
+                search = logic.get_action('datastore_search')(
+                    context,
+                    {'resource_id': resource_id}
+                )
+                count = search['total']
+            except logic.NotFound:
+                count = None
+
+            return count
+
+
         # delete the rows of the orphaned datastore tables
         delete_count = 0
         delete_error_count = 0
+
         for resource_id in resource_id_list:
+            count = datastore_record_count(context, resource_id)
+            if count is not None:
+                print('Deleting datastore table with {} entries for resource {} ...'.format(count, resource_id))
+            else:
+                print('No datastore table found for resource {}'.format(resource_id))
+                delete_error_count += 1
+                continue
+
             try:
                 logic.get_action('datastore_delete')(
                     context,
                     {'resource_id': resource_id, 'force': True}
                 )
-                print "Table '%s' deleted (not dropped)" % resource_id
-                delete_count += 1
             except AttributeError as e:
-                print("Issue with dropping resource {}".format(resource_id))
+                # datastore_delete references resource.extras when there is no resource
+                if e.message != "'NoneType' object has no attribute 'extras'":
```

Comment:
`datastore_delete` action in core ckan *always* throws an error when there is no resource related to the datastore table. However the error happens after the actual delete occurs, so it still works. I'm catching the specific error here, and raising if it's anything else at all, which will kill the cleanup script.